### PR TITLE
Phase1 (2015 + Pixel Phase1) Workflow up to Reco

### PR DESCRIPTION
**Pull request by makortel**
This PR is an update of #7625 adding fixes to tracking configuration. Original description of #7625:

```
This PR is modifying the 2017 scenario workflow to make it run till the reconstruction .
This is part of merging the upgrade code in &X cycle.

At this point it fails at the tracking step (*).
The geometry and local reco should now be validated while the reco step needs a bit more work.
Recipe : RunTheMatrix --what upgrade -l 10000

Run1/run2 workflows and code are unchanged.
```

and changes on top of that
- Bug fix in QuadrupletSeedMerger
- Typo fix in Phase1PU70_LowPtTripletStep_cff.py
- Move `clustersToSkip` parameter from TrajectoryBuilders to CkfTrackCandidateMaker in Phase1 configuration files
  - In line with #1063
  - Removes the printout `ERROR: 24BaseCkfTrajectoryBuilder has a clustersToSkip parameter set`
- Migrate Phase1 configuration files to use `refToPSet_` for PixelSeedMergerQuadruplets

The RAW2DIGI,L1Reco,RECO step of wf 10000 still fails though (in 740pre6), now with

```
in filling An exception of category 'InvalidReference' occurred.
Exception Message:
Trie::insert called with a key already in collection;
key value: ^A

304091140 1

<snip>

An exception of category 'BadAlloc' occurred while
   [0] Processing run: 1 lumi: 1 event: 1
   [1] Running path 'raw2digi_step'
   [2] Calling event method for module HcalRawToDigi/'hcalDigis'
   [3] Using EventSetup component HcalDbProducer/'hcal_db_producer' to make data HcalDbService/'' in record HcalDbRecord
   [4] Using EventSetup component HcalHardcodeCalibrations/'es_hardcode' to make data HcalRespCorrs/'' in record HcalRespCorrsRcd
Exception Message:
A std::bad_alloc exception was thrown.
The job has probably exhausted the virtual memory available to the process.
```

@threus , @venturia , @boudoul, @atricomi, @rovere, @Vinn, @veszpv, @mark-grimes : FYI

### CODE REVIEW

**Comment by venturia** [comment]
The first part of the errors reported by Matti:
Exception Message:
Trie::insert called with a key already in collection;
key value: ^A

304091140 1

is a side effect of my changes in the way multiple geometries are handled and it is harmless. We will address it later but for the moment can be ignored.

**Comment by slava77** [comment]
@cmsbuild please test

**Comment by makortel** [comment]
@venturia Thanks, good to know it is not related to the exception.

File: Geometry/CMSCommonData/python/cmsExtendedGeometry2017XML_cfi.py [hunk 1]
```diff
@@ -207,21 +206,18 @@
...
         'Geometry/EcalCommonData/data/eehier.xml',
-        'Geometry/EcalCommonData/data/ectkcable.xml',
         'Geometry/EcalCommonData/data/eealgo.xml',
         'Geometry/EcalCommonData/data/escon.xml',
         'Geometry/EcalCommonData/data/esalgo.xml',
         'Geometry/EcalCommonData/data/eeF.xml',
         'Geometry/EcalCommonData/data/eeB.xml',
         'Geometry/HcalCommonData/data/hcalrotations.xml',
-        'Geometry/HcalCommonData/data/PhaseI/hcalalgo.xml',
+        'Geometry/HcalCommonData/data/hcalalgo.xml',
```
Comment (ianna) [inline Geometry/CMSCommonData/python/cmsExtendedGeometry2017XML_cfi.py]: @boudoul - I see you are rolling back alignment with SLHC release:
https://github.com/makortel/cmssw/commit/767313c55f476f37c3539ea889de5cf610242b00#diff-9c452f0e342fb87194f2f19ceac82e3f

**Comment by ianna** [comment]
+1

**Comment by civanch** [comment]
+1

File: Geometry/CMSCommonData/python/cmsExtendedGeometry2017XML_cfi.py (hunk 1 above)
Comment (boudoul) [inline Geometry/CMSCommonData/python/cmsExtendedGeometry2017XML_cfi.py]: Hi @ianna : yes there is some hcal code not ported in 7X which prevents it to run - I will mention this today at the sw upgrade meeting

**Comment by boudoul** [comment]
+1

**Comment by slava77** [comment]
.. i'm checking it

**Comment by slava77** [comment]
I wasn't able to even start the matrix 10000 

in CMSSW_7_4_X_2015-02-12-0200+PR7645

``` runTheMatrix.py  --what upgrade -l 10000 ```

... then in step1:
```
  File "/afs/cern.ch/work/s/slava77/reltest/CMSSW_7_4_X_2015-02-12-0200-sign517/python/Configuration/Geometry/GeometryExtended2017_cff.py", line 8, in <module>
    from Geometry.TrackerNumberingBuilder.trackerNumbering2017Geometry_cfi import *
ImportError: No module named trackerNumbering2017Geometry_cfi
```

**Comment by ianna** [comment]
@slava77 and @boudoul - it should be just trackerNumberingGeometry_cfi

**Comment by makortel** [comment]
I can reproduce the failure of `runTheMatrix.py --what upgrade -l 10000` in CMSSW_7_4_X_2015-02-12-0200, but it doesn't occur in 740pre6 (I wonder what changed in between).

Anyway, the fix suggested by @ianna should be straightforward, @boudoul do you want to provide the fix or should I?

**Comment by boudoul** [comment]
hi @makortel , i'm stuck in the plenary session this morning, so please you can go ahead - If you are busy,  I can take a look beginning of this afternoon

**Comment by makortel** [comment]
There were couple of other uses of `trackerNumbering2017Geometry_cfi`, I replaced them all with `trackerNumberingGeometry_cfi`. Running the wf 10000 in CMSSW_7_4_X_2015-02-12-0200 fails now (in the first step) with
```
An exception of category 'DDException' occurred while
   [0] Processing run: 1
   [1] Calling global beginRun for module OscarProducer/'g4SimHits'
   [2] Using EventSetup component TrackerGeometricDetESModule/'trackerNumberingGeometry' to make data GeometricDet/'' in record IdealGeometryRecord
Exception Message:
DDVectorGetter: vectors are empty, cannot get array detIdShifts
```
Ideas?  @ianna @venturia @boudoul

**Comment by venturia** [comment]
Maybe Gaelle's pull request not only was not aware of Ianna's pull request that moved the numbering configuration from cfi to xml, but also of the fact that now new xml files have to be "uploaded" (sorry for the sloppy wording)

**Comment by ianna** [comment]
@makortel - please, include the following file in cmsExtendedGeometry2017XML_cfi.py:

```
'Geometry/CMSCommonData/data/PhaseI/trackerParameters.xml',
```

**Comment by makortel** [comment]
@ianna Thanks, but should it have `TrackerCommonData` instead of `CMSCommonData`? (the former seems to work but latter not)

**Comment by ianna** [comment]
@makortel - sorry, yes, it should be TrackerCommonData

**Comment by makortel** [comment]
@ianna Thanks for the confirmation.

With this fix, the wf 10000 fails again (in CMSSW_7_4_X_2015-02-12-0200) with the `HcalDbProducer`/`HcalHardcodeCalibrations`-related `std::bad_alloc` exception (same as in the description of this PR).

**Comment by ianna** [comment]
@makortel - I'm looking into it

**Comment by ianna** [comment]
@makortel - I think, you should remove Hcal from customizations in SLHCUpgradeSimulations/Configuration/python/combinedCustoms.py
```
def cust_2017(process):
    process=customisePostLS1(process)
    process=customisePhase1Tk(process)
    process=customise_HcalPhase0(process)
    process=fixRPCConditions(process)
    return process
```

**Comment by ianna** [comment]
@makortel - also RPC customization is not needed here (and it's wrong)

**Comment by ianna** [comment]
@makortel - most probably, customizations for Tracker are wrong as well as now I get a Tracker related exception:
```
----- Begin Fatal Exception 13-Feb-2015 15:13:59 CET-----------------------
An exception of category 'ProductNotFound' occurred while
   [0] Processing run: 1 lumi: 1 event: 1
   [1] Running path 'reconstruction_step'
   [2] Calling event method for module SiPixelClusterProducer/'siPixelClustersPreSplitting'
Exception Message:
Principal::getByToken: Found zero products matching all criteria
Looking for type: edm::DetSetVector<PixelDigi>
Looking for module label: siPixelDigis
Looking for productInstanceName: 

   Additional Info:
      [a] If you wish to continue processing events after a ProductNotFound exception,
add "SkipEvent = cms.untracked.vstring('ProductNotFound')" to the "options" PSet in the configuration.

----- End Fatal Exception -------------------------------------------------
Another exception was caught while trying to clean up files after the primary fatal exception.
13-Feb-2015 15:13:59 CET  Closed file file:step2.root
```

**Comment by makortel** [comment]
@ianna Thanks! I got the error in `siPixelClusterPreSplitting` too. Before starting to investigate, it looks like a clash between the tracking developments in 74X and the 2017-customise from 62X_SLHC.

**Comment by venturia** [comment]
Let me try: since we don't go through packing and unpacking for the pixel phase 1 digi, the name of the digi collection for phase 1 is slightly different.

**Comment by makortel** [comment]
@venturia `simSiPixelDigis` perhaps?

**Comment by boudoul** [comment]
The workflow defined should take into account this (was working in 62XSLHC with the right collection name ) with the cutomize ...the last error doesn't really make sense . I'll give a chance to Andrea, but I should probably start from scratch the work I did up to local reco, I suspect something is not working already from there . Then I will pull request , let it merge (hopefullY0 from that Matti can take over for reco...

**Comment by ianna** [comment]
@boudoul - there is a comment in Tracker customization in customise_Reco:
```
    # Need this line to stop error about missing siPixelDigis.
    process.MeasurementTracker.inactivePixelDetectorLabels = cms.VInputTag()
```
is it used? should it work?

**Comment by boudoul** [comment]
yes it should be there , should also work in 7X

**Comment by davidlange6** [comment]
@makortel - could you rebase this PR? - it does not merge

**Comment by makortel** [comment]
Rebased on top of HEAD of 74X (sorry for all those who want to test before the next IB). Merge conflict arose from #7724.

**Comment by ianna** [comment]
@makortel - this exception goes away if I do not remove the collection, but there will be another similar error.
```
 def customise_RawToDigi(process):
-    process.raw2digi_step.remove(process.siPixelDigis)
+    ##process.raw2digi_step.remove(process.siPixelDigis)
```
and
```
     # Need this line to stop error about missing siPixelDigis.
-    process.MeasurementTracker.inactivePixelDetectorLabels = cms.VInputTag()
+    ####process.MeasurementTracker.inactivePixelDetectorLabels = cms.VInputTag()

```

**Comment by venturia** [comment]
@makortel I have made a pull request to _your_ branch, boudoul_Phase1wF_recofixes, with a few changes that remove the need of the configuration parameter upgradeGeometry. I think that if you accept it and merge it, this pull request of yours will be updated automatically. @atricomi, can you check it? It is similar to the one I did for the pull request #7589 of @ianna . I have no idea which of the two PRs will be integrated first but I did not want to add a third one on related items.

**Comment by makortel** [comment]
@venturia Thanks, merged.

@ianna Thanks, but I'll leave the local reco to @boudoul to sort out :)

**Comment by boudoul** [comment]
Hi @makortel ! I worked with the merged version  : I managed to fix the simdigis crash  : in the customs reco I added  process.siPixelClustersPreSplitting.src="simSiPixelDigis"
[The presplitting was not existing in 62X it's something ne fin 7X, that's why it was not there by defaults in the customs]
But then I have another issue with the fake conditions that I am sorting out ... (work in progress)
I didn't PR  anything .
We can maybe have this PR already merge (unless other troubles with merging in 7X)
  Meanwhile I'm working on the local reco to fix the hopefully last 'implementation details'

**Comment by makortel** [comment]
@boudoul Ok, thanks. I would not do that change (yet) though, because the "pre-splitting clusters" are not useful as such without the cluster splitting step, that in turn requires a bit more work to integrate to Phase1 tracking configurations (e.g. now `initialStepSeeds` will fail with the missing `siPixelClusters` product). I would rather try to reproduce the 62X_SLHC-Phase1 tracking in 74X first, and once that is validated, start to add the 62X_SLHC..74X improvements one by one.

I would also let this PR merge at this point.

**Comment by makortel** [comment]
I managed to get some progress, and, since nobody has approved this PR between the last push and now, decided to push the update here. The changes are
* do not use the "pre-splitting" cluster and rechit variants (for now)
* use `pixelTracks` and `pixelVertices` (for now)
* add `initialStepSeedLayers` (bugfix, somehow I missed these year ago)

Now I get (a lot) of printouts like
```
%MSG-e SiPixelGenErrorDBObject:  SiPixelRecHitConverter:siPixelRecHits  16-Feb-2015 17:03:18 CET Run: 1 Event: 1
GenError ID for DetID 303042564 is not stored
%MSG
%MSG-e SiPixelLorentzAngle:  SiPixelRecHitConverter:siPixelRecHits  16-Feb-2015 17:03:18 CET Run: 1 Event: 1
SiPixelLorentzAngle for DetID 303042564 is not stored
%MSG

....

%MSG-e SiPixelLorentzAngle:  PixelTrackProducer:pixelTracks  16-Feb-2015 17:03:24 CET Run: 1 Event: 1
SiPixelLorentzAngle for DetID 303042564 is not stored
%MSG

....

%MSG-e SiPixelTemplateDBObject:  DuplicateTrackMerger:duplicateTrackCandidates  16-Feb-2015 17:03:24 CET Run: 1 Event: 1
Template ID for DetID 303042564 is not stored
%MSG
%MSG-e SiPixelLorentzAngle:  DuplicateTrackMerger:duplicateTrackCandidates  16-Feb-2015 17:03:24 CET Run: 1 Event: 1
SiPixelLorentzAngle for DetID 303042564 is not stored
%MSG
```
and finally the job fails with an exception
```
An exception of category 'NoRecord' occurred while
   [0] Processing run: 1 lumi: 1 event: 1
   [1] Running path 'reconstruction_step'
   [2] Calling event method for module CorrectedECALPFClusterProducer/'particleFlowClusterECAL'
Exception Message:
No "GBRDWrapperRcd" record found in the EventSetup.
 Please add an ESSource or ESProducer that delivers such a record.
```

**Comment by slava77** [comment]
Is there much benefit to have this in 74X?
We can just continue in 75X.

**Comment by davidlange6** [comment]
closing in 74x - there is the same PR open in 75x (now the development release)

**Pull request closed without merging**